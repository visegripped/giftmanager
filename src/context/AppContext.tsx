import React, { createContext, useContext } from 'react';
import { UserResponseProps } from '../util/fetchData';

export interface IAuthUpdate {
  tokenId: string;
  email: string;
}

interface IAppContext extends IAuthUpdate {
  setAuth: (valObj: IAuthUpdate) => void;
  children?: React.ReactNode;
  userId: number;
  setUserId: (id: number) => void;
  users: UserResponseProps[];
}

const defaultState = {
  tokenId: '',
  email: '',
  users: [],
  userId: 0,
  setUserId: () => {
    console.log('default context updateAuth executed');
  },
  setAuth: () => {
    console.log('default context updateAuth executed');
  },
};

export const AppContext = createContext<IAppContext>(defaultState);

export const useAppContext = () => useContext(AppContext);

export default AppContext;
