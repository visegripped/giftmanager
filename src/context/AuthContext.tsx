import {
  useState,
  createContext,
  useEffect,
  useCallback,
  PropsWithChildren,
  useContext,
} from 'react';
import { googleLogout, useGoogleLogin } from '@react-oauth/google'; // docs: https://www.npmjs.com/package/@react-oauth/google
import {
  NotificationsContext,
  AddNotificationProps,
} from '@context/NotificationsContext';

export interface AuthContextInterface {
  login: () => {};
  logout: () => {};
  accessToken: string;
  setAccessToken: (accessToken: string) => {};
}

const AuthContext = createContext(null);
// const refreshTokenUrl = import.meta.env.VITE_REFRESH_TOKEN_URL;
const googleClientId = import.meta.env.VITE_GOOGLE_CLIENT_ID;

export interface GoogleAuthResponseInterface {
  access_token: string;
  token_type: string;
  expires_in?: number;
  refresh_token?: string;
  scope: string;
}

function AuthProvider(props: PropsWithChildren) {
  // console.log(' -> props from authProvider', props);
  const [accessToken, setAccessToken] = useState(
    localStorage.getItem('access_token') || ''
  );
  const [accessTokenExpiration, setAccessTokenExpiration] = useState<
    Date | string
  >(localStorage.getItem('access_token_expiration') || '');

  const { addNotification } =
    useContext<AddNotificationProps>(NotificationsContext);

  const handleGoogleTokenExpiration = () => {
    let currentTime = new Date().getTime();
    const expirationTime = new Date(currentTime + 2 * 60 * 60 * 1000); // 2 hours for token life
    setAccessTokenExpiration(expirationTime);
    localStorage.setItem('access_token_expiration', expirationTime.toString());
  };

  const onLoginSuccess = (authResponse: GoogleAuthResponseInterface) => {
    const { access_token, refresh_token } = authResponse;
    setAccessToken(access_token);
    localStorage.setItem('access_token', access_token);
    localStorage.setItem('refresh_token', refresh_token || '');
    handleGoogleTokenExpiration();
  };

  const onLoginFailure = (authResponse: GoogleAuthResponseInterface) => {
    addNotification({
      message: `Something has gone wrong with your authentication.  This may help: ${JSON.stringify(
        authResponse
      )}`,
      type: 'error',
    });
  };

  const login = useGoogleLogin({
    onSuccess: onLoginSuccess,
    onError: onLoginFailure,
    client_id: googleClientId,
    scope: 'openid profile email',
  });

  const logout = () => {
    localStorage.removeItem('access_token');
    localStorage.removeItem('tokenId');
    setAccessToken('');
    googleLogout();
  };

  const tokenIsValid = async () => {
    if (!accessToken) {
      return false;
    }
    const currentTime = new Date();
    if (currentTime >= accessTokenExpiration) {
      return false;
    }
    if (!validateTokenViaAPI()) {
      return false;
    }

    return true;
  };

  const validateTokenViaAPI = useCallback(async () => {
    if (!accessToken) return false;

    try {
      const response = await fetch(
        'https://www.googleapis.com/oauth2/v3/tokeninfo?access_token=' +
        accessToken
      );
      const data = await response.json();

      if (data.error_description) {
        logout();
        return false;
      }

      return true;
    } catch (error) {
      logout();
      return false;
    }
  }, [accessToken, accessTokenExpiration]);

  // const validateAndUpdateUser = (profile: userProfileInterface) => {
  //   if (profile.email) { }
  //   else {
  //     Logout();
  //   }
  // }

  useEffect(() => {
    if (!accessToken || !tokenIsValid()) {
      logout();
    }
  }, [accessToken]);

  return (
    <AuthContext.Provider
      value={{ login, logout, accessToken, setAccessToken }}
    >
      {props.children}
    </AuthContext.Provider>
  );
}

export { AuthContext, AuthProvider };
export default AuthContext;
