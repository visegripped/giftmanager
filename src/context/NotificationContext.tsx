// https://felixgerschau.com/react-typescript-context/
import { createContext, useContext } from 'react';

export interface IMessage {
  type: 'error' | 'warn' | 'info';
  report: string;
  id?: string;
  meta?: object;
}

interface INotificationContext {
  messages: IMessage[];
  removeMessage: (messageId: string) => void;
  addMessage: (message: IMessage) => void;
}

const defaultState = {
  messages: [],
  addMessage: (message: IMessage) => {
    console.log('default addMessage triggered.', message.report);
  },
  removeMessage: (messageId: string) => {
    console.log('default removeMessage triggered.', messageId);
  },
};

export const NotificationContext = createContext<INotificationContext>(defaultState);

export const useNotificationContext = () => useContext(NotificationContext);

export default NotificationContext;
