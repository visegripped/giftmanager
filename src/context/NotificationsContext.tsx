import {
  useState,
  createContext,
  PropsWithChildren,
  useCallback,
  useMemo,
} from 'react';
import { UUID } from 'uuidjs';

export type AddNotificationTypeProp = 'error' | 'success' | 'info' | 'warn';

export interface AddNotificationProps {
  type: AddNotificationTypeProp;
  message: string;
  persist?: boolean | null;
  clearDuration?: number;
}

export interface NotificationProps extends AddNotificationProps {
  uuid: string | number;
}

export interface NotificationContextProps {
  notifications: { [k: string]: NotificationProps };
  setNotifications: () => {};
  addNotification: ({}: AddNotificationProps) => {};
  removeNotification: (u: string | number) => {};
}

const NotificationsContext = createContext({});

const NotificationsProvider = (props: PropsWithChildren) => {
  const [notifications, setNotifications] = useState<{ [k: string]: NotificationProps }>({});

  const removeNotification = useCallback((uuid: string | number) => {
    setNotifications((prevNotifications) => {
      const updatedNotifications = { ...prevNotifications };
      delete updatedNotifications[uuid];
      return updatedNotifications;
    });
  }, []);

  const addNotification = useCallback((notificationObj: NotificationProps) => {
    const uuid = UUID.generate();
    notificationObj.uuid = uuid; // add to the individual notification for easy lookup later.
    setNotifications((prevNotifications) => ({
      ...prevNotifications,
      [uuid]: notificationObj,
    }));
  }, []);

  const contextValue = useMemo(
    () => ({
      notifications,
      setNotifications,
      addNotification,
      removeNotification,
    }),
    [notifications, addNotification, removeNotification]
  );

  return (
    <NotificationsContext.Provider value={contextValue}>
      {props.children}
    </NotificationsContext.Provider>
  );
};

export { NotificationsContext, NotificationsProvider };
export default NotificationsContext;
