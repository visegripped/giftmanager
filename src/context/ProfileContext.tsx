import { useState, createContext, useMemo } from 'react';
import { UserProfileInterface } from '../types/types';

export interface ProfileContextInterface {
  myProfile: UserProfileInterface;
  setMyProfile: (a: object) => {};
  fetchGoogleProfile: () => {};
}

const ProfileContext = createContext({});

const ProfileProvider = (props: React.PropsWithChildren) => {
  const [myProfile, setMyProfile] = useState({});

  const contextValue = useMemo(
    () => ({
      myProfile,
      setMyProfile,
    }),
    [myProfile]
  );

  return (
    <ProfileContext.Provider value={contextValue}>
      {props.children}
    </ProfileContext.Provider>
  );
};

export { ProfileContext, ProfileProvider };
export default ProfileContext;
