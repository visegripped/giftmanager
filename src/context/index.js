export { default as AuthContext, AuthProvider } from './AuthContext';
export {
  default as NotificationsContext,
  NotificationsProvider,
} from './NotificationsContext';
export {
  default as ProfileContext,
  ProfileProvider,
  ProfileContextInterface,
} from './ProfileContext';
