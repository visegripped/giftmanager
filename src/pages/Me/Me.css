.table-container {
  height: 90%;
  padding: var(--spacing);
}

.ag-root-wrapper {
  opacity: 80%;
}

.row-removed {
  background-color: var(--container-bg-warning);
}
