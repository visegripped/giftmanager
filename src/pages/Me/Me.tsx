import { useState, useEffect } from 'react';
import Icon from '@components/Icon';
import { AgGridReact } from 'ag-grid-react'; // React Data Grid Component
import 'ag-grid-community/styles/ag-grid.css'; // Mandatory CSS required by the Data Grid
import 'ag-grid-community/styles/ag-theme-quartz.css'; // Optional Theme applied to the Data Grid
import './Me.css';
import fetchData from '@utilities/fetchData';


/*
To do:
  Status column: 
      if item is removed, row should turn warning/error color.
      remove button needs to work
      edit button should allows changing the link and description, NOT the name.
  API response should be cleaned up to only return data that doesn't reveal anything.
  Page needs to be responsive, for smaller browsers, left col should be on top.
  Need an Add item section
    includes: name, link, description (rename note)
    updates DB
    refetches list.
    updates table.
*/

type propsInterface = {
  userid: string | undefined;
  date_added: number;
  name: string;
  note: string;
  date_received: number;
  status: null | 'nochange' | 'uncancel' | 'remove' | 'purchased' | 'reserved';
  removed: 1 | 0;
  link: string;
  giftid: number;
  groupid: number;
};

type myItemListInterface = {
  myItemList: propsInterface[];
};

type tableDataInterface = {
  data: propsInterface;
};

const Link = (props: propsInterface) => {
  const { link, name } = props;
  return (
    <a href={link} target="_blank">
      {name}
    </a>
  );
};
const StatusDD = (props: propsInterface) => {
  const { removed } = props.data;
  return (
    <>
      {removed === 1 ? (
        <Icon icon='plus' />
      ) : (
        <>
          <Icon icon='edit' />
          <Icon icon='delete' />
        </>
      )}
    </>
  );
};

const linkedName = (props: tableDataInterface) => {
  return <Link {...props.data} />;
};

// const adjustedStatus = (props: tableDataInterface) => { };

// change status title to actions: add a remove button.
const Table = (props: myItemListInterface) => {
  const { myItemList } = props;
  const [colDefs] = useState([
    { field: 'name', sortable: true, cellRenderer: linkedName, sort: 'asc' },
    { field: 'description' },
    {
      field: 'removed',
      cellRenderer: StatusDD,
      headerName: 'actions',
      flex: 2,
    },
  ]);

  const rowClassRules = {
    'row-removed': 'data.removed >= 1',
  };

  return (
    <>
      <AgGridReact
        rowData={myItemList}
        columnDefs={colDefs}
        rowClassRules={rowClassRules}
        style={{ width: '100%', height: '100%' }}
      />
    </>
  );
};

const Me = () => {
  const [myItemList, setMyItemList] = useState([]);

  // const [userProfile] = useState({ }); //, setUserProfile

  useEffect(() => {
    const response = fetchData({
      task: 'getMyList',
      myuserid: 1,
      userid: 1,
    });
    response &&
      response.then((data) => {
        setMyItemList(data.success);
      });
  }, []);

  return (
    <>
      <h2 className="page-heading">YOURNAME's List</h2>
      <section className="table-container ag-theme-quartz responsive-grid-container responsive-grid-columns responsive-grid-sidebar">

        <form className='form'>
          <fieldset className='fieldset'>
            <legend className='legend'>Add item</legend>

            <label className='label' >Name</label>
            <div className='input-container'><input type="text" name="name" /></div>


            <label>URL</label>
            <div className='input-container'><input type="url" name="link" /></div>


            <label>Description</label>
            <div className='input-container'><textarea name="description"></textarea></div>

            {/* <button><img src={iconPlus} alt="" /> Add item</button> */}
          </fieldset>
        </form>

        <>
          {myItemList.length ? (
            <Table myItemList={myItemList} />
          ) : (
            <h3>Fetching data...</h3>
          )}
        </>

      </section>
    </>
  );
};

export default Me;
