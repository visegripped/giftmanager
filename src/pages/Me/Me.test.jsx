import Me from './Me';
import { render } from '@testing-library/react';
import { ProfileProvider } from '../../context/ProfileContext';
import { NotificationsProvider } from '../../context/NotificationsContext';

describe('The Me page', () => {
  describe('when no user context is provided', () => {
    it('should match snapshot', () => {
      const { container } = render(
        <ProfileProvider>
          <NotificationsProvider>
            <Me />
          </NotificationsProvider>
        </ProfileProvider>
      );
      expect(container).toMatchSnapshot();
    });
  });
});
