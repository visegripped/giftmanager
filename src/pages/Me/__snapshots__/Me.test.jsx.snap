// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`The Me page > when no user context is provided > should match snapshot 1`] = `
<div>
  <h2
    class="page-heading"
  >
    Welcome to your list.
  </h2>
  <section
    class="table-container ag-theme-quartz responsive-grid-container responsive-grid-columns responsive-grid-sidebar"
  >
    <form
      class="form"
    >
      <fieldset
        class="fieldset"
      >
        <legend
          class="legend"
        >
          Add item to my list
        </legend>
        <label
          class="label"
        >
          Name
        </label>
        <div
          class="input-container"
        >
          <input
            name="name"
            required=""
            type="text"
            value=""
          />
        </div>
        <label>
          Link
        </label>
        <div
          class="input-container"
        >
          <input
            name="link"
            type="url"
            value=""
          />
        </div>
        <label>
          Description
        </label>
        <div
          class="input-container"
        >
          <textarea
            name="description"
          />
        </div>
        <button
          class="button button--undefined"
          type="submit"
        >
          <span
            class="icon"
          >
            <svg
              fill="none"
              height="20"
              viewBox="0 0 24 24"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M17.983 11.0265L12.983 11.0195L12.99 6.01953C12.99 5.75431 12.8846 5.49996 12.6971 5.31242C12.5095 5.12489 12.2552 5.01953 11.99 5.01953C11.7248 5.01953 11.4704 5.12489 11.2829 5.31242C11.0953 5.49996 10.99 5.75431 10.99 6.01953L10.983 11.0195L5.98297 11.0115C5.71775 11.0115 5.4634 11.1169 5.27586 11.3044C5.08833 11.492 4.98297 11.7463 4.98297 12.0115C4.98297 12.2767 5.08833 12.5311 5.27586 12.7186C5.4634 12.9062 5.71775 13.0115 5.98297 13.0115L10.983 13.0195L10.975 18.0195C10.975 18.2847 11.0803 18.5391 11.2679 18.7266C11.4554 18.9142 11.7098 19.0195 11.975 19.0195C12.2402 19.0195 12.4945 18.9142 12.6821 18.7266C12.8696 18.5391 12.975 18.2847 12.975 18.0195L12.983 13.0195L17.983 13.0265C18.2482 13.0265 18.5025 12.9212 18.6901 12.7336C18.8776 12.5461 18.983 12.2917 18.983 12.0265C18.983 11.7613 18.8776 11.507 18.6901 11.3194C18.5025 11.1319 18.2482 11.0265 17.983 11.0265Z"
                fill="#FFF"
              />
            </svg>
          </span>
          <span>
            Add
          </span>
        </button>
      </fieldset>
    </form>
    <h3>
      There are no items in this your list. Please add some.
    </h3>
  </section>
</div>
`;
